expression/builtin_cast.go

```diff
@@ -1804,6 +1804,12 @@ func BuildCastFunction4Union(ctx sessionctx.Context, expr Expression, tp *types.
 
 // BuildCastFunction builds a CAST ScalarFunction from the Expression.
 func BuildCastFunction(ctx sessionctx.Context, expr Expression, tp *types.FieldType) (res Expression) {
+	if res, success := TryPushCastDownToControlFunctionForHybridType(ctx, expr, tp); success {
+		if tp.EvalType() != types.ETJson {
+			res = FoldConstant(res)
+		}
+		return res
```

Comment:
`BuildCastFunc` will also be used for explicit `cast`, if we return here, we may build a wrong scalar function.
e.g. `select cast(if ()) from t ;`